edb/edgeql/compiler/casts.py

```diff
@@ -540,6 +553,166 @@ def _cast_tuple(
+        new_stype: s_types.Range,
+        cardinality_mod: Optional[qlast.CardinalityModifier],
+        *,
+        srcctx: Optional[parsing.ParserContext],
+        ctx: context.ContextLevel) -> irast.Set:
+
+    with ctx.new() as subctx:
+        subctx.anchors = subctx.anchors.copy()
+        source_path = subctx.create_anchor(ir_set, 'a')
+        check = qlast.FunctionCall(
+            func=('__std__', '__range_validate_json'), args=[source_path]
+        )
+        check_ir = dispatch.compile(check, ctx=subctx)
+        source_path = subctx.create_anchor(check_ir, 'b')
+
+        range_el_t = new_stype.get_element_type(ctx.env.schema)
+        ql_range_el_t = typegen.type_to_ql_typeref(range_el_t, ctx=subctx)
+        bool_t = ctx.env.get_track_schema_type(sn.QualName('std', 'bool'))
+        ql_bool_t = typegen.type_to_ql_typeref(bool_t, ctx=subctx)
+
+        cast = qlast.FunctionCall(
+            func=('__std__', 'range'),
+            args=[
+                qlast.TypeCast(
+                    expr=qlast.FunctionCall(
+                        func=('__std__', 'json_get'),
+                        args=[
+                            source_path,
+                            qlast.StringConstant(value='lower'),
+                        ],
+                    ),
+                    type=ql_range_el_t,
+                ),
+                qlast.TypeCast(
+                    expr=qlast.FunctionCall(
+                        func=('__std__', 'json_get'),
+                        args=[
+                            source_path,
+                            qlast.StringConstant(value='upper'),
+                        ],
+                    ),
+                    type=ql_range_el_t,
+                ),
+            ],
+            kwargs={
+                "inc_lower": qlast.TypeCast(
+                    expr=qlast.FunctionCall(
+                        func=('__std__', 'json_get'),
+                        args=[
+                            source_path,
+                            qlast.StringConstant(value='inc_lower'),
+                        ],
+                    ),
+                    type=ql_bool_t
+                ),
+                "inc_upper": qlast.TypeCast(
+                    expr=qlast.FunctionCall(
+                        func=('__std__', 'json_get'),
+                        args=[
+                            source_path,
+                            qlast.StringConstant(value='inc_upper'),
+                        ],
+                    ),
+                    type=ql_bool_t
+                ),
+                "empty": qlast.TypeCast(
+                    expr=qlast.FunctionCall(
+                        func=('__std__', 'json_get'),
+                        args=[
+                            source_path,
+                            qlast.StringConstant(value='empty'),
+                        ],
+                    ),
+                    type=ql_bool_t
+                ),
```

Comment:
If I read this correctly this should cause the `empty` argument to be an `{}` if the input JSON doesn't have an explicit `empty` field. Which, in turn, should cause the entire `range` call to result in an `{}` as well, because the named argument `empty` is not optional, it merely has a default and _can be omitted_ entirely.

Reply:
These specializations are somewhat different from the normal compilation semantics.  The range constructor is implemented as a non-strict SQL function, so it _will_ get called on `NULL` input to even a non-optional argument (if called directly from a specialization like here), and the implementation uses `empty` in a way that tolerates nulls.

Reply:
Would it be better to pass False in to `json_get` explicitly and avoid any funny business?

Reply:
@vpetrovykh was right that this is sketchy. The non-strictness of the range constructor is not at all limited to being called directly from this kind of specialization:
```
edgedb> select range(<int64>{}, empty:=<optional bool>$0);
Parameter <bool>$0 (Ctrl+D for empty set `{}`): 
[{"inc_lower": false, "inc_upper": false, "lower": null, "upper": null}]
edgedb> select range(<int64>{}, inc_upper:=<optional bool>$0);
Parameter <bool>$0 (Ctrl+D for empty set `{}`): 
edgedb error: InvalidValueError: range constructor flags argument must not be null
```
as one example.
Both examples ought to return an empty set, since they pass an empty set to a non-optional argument, but one returns a value and the other raises an error.